Add FirstEnabledTraitOrDefault helper method.

RoosterDragon opened the pull request:
This avoids the allocations caused by LINQ when using traits.FirstOrDefault(Exts.IsTraitEnabled). This is important in FrozenActorLayer.RefreshState which is called very often. We apply the new helper method to all areas using the old pattern. An overload that takes an array allows arrays to be enumerated without causing allocations.

RoosterDragon committed:
Add FirstEnabledTraitOrDefault helper method.

This avoids the allocations caused by LINQ when using traits.FirstOrDefault(Exts.IsTraitEnabled). This is important in FrozenActorLayer.RefreshState which is called very often. We apply the new helper method to all areas using the old pattern. An overload that takes an array allows arrays to be enumerated without causing allocations.

RoosterDragon commented on OpenRA.Mods.Common/Lint/CheckTooltips.cs:
```diff
@@ -26,7 +26,7 @@ public void Run(Action<string> emitError, Action<string> emitWarning, Ruleset ru
 					continue;
 
-				var tooltip = actorInfo.Value.TraitInfos<TooltipInfo>().FirstOrDefault(Exts.IsTraitEnabled);
+				var tooltip = actorInfo.Value.TraitInfos<TooltipInfo>().FirstEnabledTraitOrDefault();
```
By the way, does `Exts.IsTraitEnabled` work on `TraitInfo`s? I figured it was just for traits (hoping to learn something :))

pchote replied:
It doesn't. This is a copy-pasta error of an older logical error.  Lets get rid of it.

pchote replied:
You can use `ConditionalTraitInfo.EnabledByDefault` instead.

RoosterDragon replied:
A cursory glance shows places other than just tooltips affected, I have created #14445 to track this instead.

reaperrr replied:
Wouldn't it still be better to fix the places that are touched by this PR, though?

The old bogus look-up would just ignore the Exts.IsTraitsEnabled, but now this uses trait-specific look-up for TraitInfo, will this even work (without causing regressions vs. bleed, I mean)?

Besides, I couldn't find any other places (or rather, the only other similar place I could find already uses EnabledByDefault as @pchote suggested), though it's possible my search approach didn't catch them all.

reaperrr commented:
@RoosterDragon I suspect we have quite a few cases of `traits.FirstOrDefault(t => !t.IsTraitDisabled)` as well, would it win us anything to make those use this helper method, too?

RoosterDragon committed:
Fix bad uses of FirstEnabledTraitOrDefault on TraitInfos.

These are not traits so this method does not work. We can use EnabledByDefault on the ConditionalTraitInfo instead.

RoosterDragon commented:
Have fixed bad `Exts.IsTraitEnabled` calls affected by this PR. Remaining bad uses are covered in #14445.

pchote commented on OpenRA.Game/Exts.cs:
```diff
@@ -503,6 +503,26 @@ public static bool IsTraitEnabled<T>(T t)
+		public static T FirstEnabledTraitOrDefault<T>(this T[] ts)
+		{
+			// PERF: Avoid LINQ.
+			foreach (var t in ts)
```
General question: Is the compiler smart enough to generate code to enumerate these by index (`for (var i = 0; i < ts.Length; i++)`, or does it allocate an enumerator object?

pchote replied:
According to the internet: yes.

RoosterDragon replied:
Yes, the compiler special cases `foreach` on arrays to do exactly that.

pchote reviewed (approved):
LGTM, but i'm interested in finding out the answer to my question above before hitting the merge button.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5d8c76f4d51f7c75fe38bed47b3bf2c70e6493f0
